src/MemoryTracing/MemoryTracingUtils.cpp

```diff
@@ -85,4 +88,137 @@ std::optional<SystemMemoryUsage> GetSystemMemoryUsage() noexcept {
   return parsing_result;
 }
 
+int64_t GetVmRssFromProcessStatus(std::string_view status_content) {
+  if (status_content.empty()) return kMissingInfo;
+
+  std::vector<std::string> lines = absl::StrSplit(status_content, '\n');
+  if (lines.empty()) return kMissingInfo;
+
+  for (std::string_view line : lines) {
+    if (absl::StartsWith(line, "VmRSS")) {
+      std::vector<std::string> splits = absl::StrSplit(line, ' ', absl::SkipWhitespace{});
+      // According to the kernel code
+      // https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/fs/proc/task_mmu.c,
+      // the size unit of the VmRSS field in the /proc/<PID>/status file is fixed to "kB", which
+      // implies 1024 Bytes. We report in "KB" in ProcessSystemUsage.
+      CHECK(splits.size() == 3 && splits[2] == "kB");
+      int64_t rss_kb;
+      if (absl::SimpleAtoi(splits[1], &rss_kb)) return rss_kb;
+      return kMissingInfo;
+    }
+  }
+
+  return kMissingInfo;
+}
+
+std::optional<ProcessMemoryUsage> GetProcessMemoryUsage(uint32_t pid) noexcept {
+  uint64_t current_timestamp_ns = orbit_base::CaptureTimestampNs();
+
+  // Extract the VmRSS value in kilobytes from the /proc/<pid>/status file.
+  ErrorMessageOr<std::string> reading_result =
+      orbit_base::ReadFileToString(absl::StrFormat("/proc/%d/status", pid));
+  if (!reading_result.has_value()) {
+    ERROR("%s", reading_result.error().message());
```

Comment:
Since you can get an error here, return ErrorMessageOr instead of optional?

Reply:
Good idea! Changed to return `ErrorMessageOr` in `GetCGroupMemoryUsage` and `GetProcessMemoryUsage`.
Also add one more commit to change the `GetSystemMemoryUsage` to return `ErrorMessageOr`.

Thanks!